runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -4788,341 +4800,332 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
-      iComment("Init 2st dim size field.");
-   if (TR::Compiler->om.compressObjectReferences())
+   else
       {
-      // Clear padding in Discontiguous array header layout. Since size field is 4 bytes wide, adding 4 to size field gets us to the padding.
-      cursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, secondDimLenReg, generateS390MemoryReference(temp2Reg, fej9->getOffsetOfDiscontiguousArraySizeField() + 4, cg));
-      iComment("Clear padding.");
-      }
-
-#if defined(J9VM_GC_SPARSE_HEAP_ALLOCATION)
-      if (isOffHeapAllocationEnabled)
-         {
-         // Since 2nd dimension array is 0 length, we can use secondDimLenReg
-         // to clear dataAddr field. secondDimLenReg is expected to be NULL
-         // at this point.
-         cursor = generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, secondDimLenReg, generateS390MemoryReference(temp2Reg, fej9->getOffsetOfDiscontiguousDataAddrField(), cg));
-         iComment("Clear 2nd dim dataAddr field.");
-         }
-#endif /* J9VM_GC_SPARSE_HEAP_ALLOCATION */
+      // Store the compressed 32 bit leaf address in the relevant element of the first dim array.
+      cursor = generateRXInstruction(cg, TR::InstOpCode::STFH, node, scratchReg, generateS390MemoryReference(sizeReg, 0, cg), cursor);
       }
 
-   if (NULL == temp3Reg)
-      temp3Reg = cg->allocateRegister();
-
-   // Store 2nd dim element into 1st dim array slot, compress temp2 if needed
-   if (comp->target().is64Bit() && comp->useCompressedPointers())
+   // Load the next leaf address in dim1SizeReg.
+   cursor = generateRREInstruction(cg, TR::InstOpCode::ALGFR, node, dim1SizeReg, dim2SizeReg, cursor);
+   // Load the next element address of the first dim array in sizeReg.
+   cursor = generateRXInstruction(cg, TR::InstOpCode::LA, node, sizeReg, generateS390MemoryReference(sizeReg, elementSize, cg), cursor);
+   if(shiftAmount > 0)
       {
-      int32_t shiftAmount = TR::Compiler->om.compressedReferenceShift();
-      generateRRInstruction(cg, TR::InstOpCode::LGR, node, temp3Reg, temp2Reg);
-      if (shiftAmount != 0)
-         {
-         generateRSInstruction(cg, TR::InstOpCode::SRAG, node, temp3Reg, temp3Reg, shiftAmount);
-         }
-      generateRXInstruction(cg, TR::InstOpCode::ST, node, temp3Reg, generateS390MemoryReference(temp1Reg, 0, cg));
+      // Load the next compressed leaf address in high 32 bits of scratchReg.
+      cursor = generateRRFInstruction(cg, TR::InstOpCode::ALHHHR, node, scratchReg, scratchReg, dim2SizeReg, cursor);
       }
-   else
+   cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, scratchReg, secondDimLabel, cursor);
+   cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_B, node, controlFlowEndLabel, cursor);
+   iComment("Allocation done!");
+
+   /********************************************* Unreachable zone *********************************************/
```

Comment:
What do you mean by unreachable zone ?

Reply:
Ok, so I do not like having two `XC` instruction to perform the clearing and having EXRL separate target. We should be able to generate better loop for that. In any case, this comment is not that clear. Can we rewrite the loop so that we do not need extra XC to be generated ?